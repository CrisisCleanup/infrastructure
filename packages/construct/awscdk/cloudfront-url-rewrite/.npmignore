# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
/.projen/
/test-reports/
junit.xml
/coverage/
/test/
/src/
!/dist/
!/dist/**/*.js
!/dist/**/*.d.ts
dist
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.eslintrc.json
!.jsii
/tsconfig.json
/tsconfig.dev.json
