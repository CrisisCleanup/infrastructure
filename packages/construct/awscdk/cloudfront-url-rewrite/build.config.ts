// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

import { compositePreset } from '@arroyodev-llc/utils.unbuild-composite-preset'
import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
	name: 'construct.awscdk.cloudfront-url-rewrite',
	declaration: false,
	clean: true,
	entries: [
		'./src/index',
		{
			input: './src/index',
			declaration: true,
		},
		{
			input: './src/handler.function',
			declaration: false,
		},
	],
	sourcemap: true,
	rollup: {
		emitCJS: true,
		cjsBridge: true,
		esbuild: {
			treeShaking: true,
			sourcemap: true,
		},
	},
	preset: compositePreset(),
	failOnWarn: false,
})
