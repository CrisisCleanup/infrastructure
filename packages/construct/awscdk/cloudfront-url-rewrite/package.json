{
  "name": "@crisiscleanup/construct.awscdk.cloudfront-url-rewrite",
  "repository": {
    "type": "git",
    "url": "https://github.com/CrisisCleanup/infrastructure"
  },
  "scripts": {
    "build": "npx projen build",
    "clean": "pnpm exec projen clean",
    "compat": "npx projen compat",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "docgen": "npx projen docgen",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "package-all": "npx projen package-all",
    "package:js": "npx projen package:js",
    "post-compile": "npx projen post-compile",
    "post-install": "npx projen post-install",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "stub": "npx projen stub",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch"
  },
  "author": {
    "name": "CrisisCleanup",
    "url": "https://crisiscleanup.org",
    "organization": true
  },
  "devDependencies": {
    "@arroyodev-llc/utils.unbuild-composite-preset": "^0.1.3",
    "@types/aws-lambda": "^8.10.122",
    "@types/jest": "^29.5.11",
    "@types/node": "^18",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "aws-cdk-lib": "2.97.0",
    "constructs": "10.2.70",
    "eslint": "^8",
    "eslint-config-prettier": "^9.0.0",
    "eslint-import-resolver-node": "^0.3.9",
    "eslint-import-resolver-typescript": "^3.6.1",
    "eslint-plugin-import": "^2.28.1",
    "eslint-plugin-prettier": "^5.0.0",
    "jest": "^29.7.0",
    "jest-junit": "^15",
    "jsii": "~5",
    "jsii-diff": "^1.89.0",
    "jsii-docgen": "^9.1.2",
    "jsii-pacmak": "^1.89.0",
    "jsii-rosetta": "~5",
    "lint-staged": "^14.0.1",
    "npm-check-updates": "^16",
    "prettier": "^3.0.3",
    "projen": "^0.73.31",
    "ts-jest": "^29.1.2",
    "typescript": "~5.1",
    "unbuild": "^2.0.0"
  },
  "peerDependencies": {
    "aws-cdk-lib": "^2.97.0",
    "constructs": "^10.2.70"
  },
  "dependencies": {
    "zod": "^3.22.2"
  },
  "keywords": [
    "cdk"
  ],
  "engines": {
    "node": ">= 18.16.0"
  },
  "license": "Apache-2.0",
  "version": "0.0.0",
  "jest": {
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)"
    ],
    "clearMocks": true,
    "collectCoverage": true,
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text"
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "watchPathIgnorePatterns": [
      "/node_modules/"
    ],
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports"
        }
      ]
    ],
    "preset": "ts-jest",
    "globals": {
      "ts-jest": {
        "tsconfig": "tsconfig.dev.json"
      }
    }
  },
  "types": "./dist/index.d.ts",
  "stability": "stable",
  "jsii": {
    "outdir": "dist",
    "targets": {},
    "tsc": {
      "outDir": "dist",
      "rootDir": "src"
    }
  },
  "type": "module",
  "sideEffects": true,
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist"
  ],
  "packageManager": "pnpm@8.7.6",
  "lint-staged": {
    "*.{ts,tsx,mts,cts,js,jsx,jts,jts}": [
      "pnpm exec eslint --fix --no-error-on-unmatched-pattern"
    ],
    "*.{yaml,yml}": [
      "pnpm exec prettier --write"
    ]
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
