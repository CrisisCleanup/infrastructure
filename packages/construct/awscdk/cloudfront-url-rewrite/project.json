{
  "name": "construct.awscdk.cloudfront-url-rewrite",
  "root": "packages/construct/awscdk/cloudfront-url-rewrite",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/assets"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "post-upgrade": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-upgrade",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "upgrade": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen upgrade",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "package:js": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:js",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "stub": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen stub",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "post-install": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-install",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    },
    "clean": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen clean",
        "cwd": "packages/construct/awscdk/cloudfront-url-rewrite"
      }
    }
  },
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"npx projen\"."
}
