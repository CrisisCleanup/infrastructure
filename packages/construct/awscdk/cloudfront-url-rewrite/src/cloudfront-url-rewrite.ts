import * as cloudfront from 'aws-cdk-lib/aws-cloudfront'
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins'
import { Construct } from 'constructs'
import {
	HandlerLambdaCloudFrontFunction,
	type HandlerLambdaCloudFrontFunctionProps,
} from './handler.cfn-function.ts'

/**
 * Represents the properties required to configure URL rewriting for CloudFront.
 */
export interface CloudFrontUrlRewriteProps {
	/**
	 * CloudFront distribution to which the URL rewrite will be applied.
	 */
	readonly distribution: cloudfront.Distribution
	/**
	 * Optional origin to use for the behavior.
	 * @default - {@link cloudfront.IOrigin} with {@link fromHostname}
	 */
	readonly origin?: cloudfront.IOrigin
	/**
	 * Optional path for the behavior.
	 * @default - '/*'
	 */
	readonly behaviorPath?: string
	/**
	 * The hostname to rewrite from.
	 */
	readonly fromHostname: string
	/**
	 * The hostname to rewrite to.
	 */
	readonly toHostname: string
	/**
	 * The redirect URI match pattern used for substitution.
	 */
	readonly redirectUriPattern: string
	/**
	 * Target URI pattern for substitution.
	 */
	readonly targetUriPattern: string
	/**
	 * Optional props for the function.
	 */
	readonly functionProps?: Partial<HandlerLambdaCloudFrontFunctionProps>
}

export class CloudFrontUrlRewrite extends Construct {
	readonly distribution: cloudfront.Distribution
	readonly origin: cloudfront.IOrigin

	constructor(scope: Construct, id: string, props: CloudFrontUrlRewriteProps) {
		super(scope, id)
		this.distribution = props.distribution

		const handler = new HandlerLambdaCloudFrontFunction(this, 'Handler', {
			...(props.functionProps ?? {}),
			fromHostname: props.fromHostname,
			redirectUriPattern: props.redirectUriPattern,
			targetUriPattern: props.targetUriPattern,
			toHostname: props.toHostname,
		})

		this.origin = props.origin ?? new origins.HttpOrigin(props.fromHostname)
		this.distribution.addBehavior(props.behaviorPath ?? '/*', this.origin, {
			functionAssociations: [
				{
					function: handler,
					eventType: cloudfront.FunctionEventType.VIEWER_REQUEST,
				},
			],
		})
	}
}
