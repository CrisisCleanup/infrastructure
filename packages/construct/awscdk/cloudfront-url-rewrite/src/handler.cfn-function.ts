import fs from 'node:fs'
import * as path from 'path'
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront'
import { type Construct } from 'constructs'

/**
 * Props for HandlerLambdaCloudFrontFunction
 */
export interface HandlerLambdaCloudFrontFunctionProps
	extends Omit<cloudfront.FunctionProps, 'code'> {
	/**
	 * The hostname to rewrite from.
	 */
	readonly fromHostname: string
	/**
	 * The hostname to rewrite to.
	 */
	readonly toHostname: string
	/**
	 * The redirect URI match pattern used for substitution.
	 */
	readonly redirectUriPattern: string
	/**
	 * Target URI pattern for substitution.
	 */
	readonly targetUriPattern: string
}

/**
 * An AWS CloudFront function which executes src/handler.lambda.ts.
 */
export class HandlerLambdaCloudFrontFunction extends cloudfront.Function {
	constructor(
		scope: Construct,
		id: string,
		props: HandlerLambdaCloudFrontFunctionProps,
	) {
		const filePath = path.join(__dirname, '../dist/handler.function.mjs')
		const content = fs
			.readFileSync(filePath, 'utf8')
			.replace('<FROM_HOSTNAME>', props.fromHostname)
			.replace('<TO_HOSTNAME>', props.toHostname)
			.replace('<REDIRECT_URI_PATTERN>', props.redirectUriPattern)
			.replace('<TARGET_URI_PATTERN>', props.targetUriPattern)
		super(scope, id, {
			comment: 'src/handler.lambda.ts',
			...props,
			code: cloudfront.FunctionCode.fromInline(content),
		})
	}
}
