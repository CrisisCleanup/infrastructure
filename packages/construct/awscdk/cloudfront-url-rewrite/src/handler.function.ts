/* eslint-disable no-var */
import type { CloudFrontFunctionsEvent } from 'aws-lambda'

var FROM_HOSTNAME = '<FROM_HOSTNAME>'
var REDIRECT_URI_PATTERN = new RegExp('<REDIRECT_URI_PATTERN>', 'g')
var TO_HOSTNAME = '<TO_HOSTNAME>'
var TARGET_URI_PATTERN = '<TARGET_URI_PATTERN>'

/**
 * cloudfront-js supports a limited subset of javascript/ecma features.
 */
// eslint-disable-next-line @typescript-eslint/require-await,@typescript-eslint/no-unused-vars
function handler(event: CloudFrontFunctionsEvent) {
	var request = event.request
	var headers = request.headers
	var uri = request.uri

	var hostParams = headers.host
	var host = ''
	if (hostParams && hostParams.value) {
		host = hostParams.value
	}

	if (!host || host !== FROM_HOSTNAME) {
		return request
	}

	var newUri = uri.replace(REDIRECT_URI_PATTERN, TARGET_URI_PATTERN)
	var newUrl = `https://${TO_HOSTNAME}${newUri}`
	return {
		statusCode: 301,
		statusDescription: 'Moved Permanently',
		headers: {
			location: { value: newUrl },
		},
	}
}

export default handler
