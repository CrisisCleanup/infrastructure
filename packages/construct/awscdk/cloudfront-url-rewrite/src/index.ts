export * from './cloudfront-url-rewrite'
export {
	type HandlerLambdaCloudFrontFunctionProps as CloudFrontRewriteFunctionProps,
	HandlerLambdaCloudFrontFunction as CloudFrontRewriteFunction,
} from './handler.cfn-function.ts'
