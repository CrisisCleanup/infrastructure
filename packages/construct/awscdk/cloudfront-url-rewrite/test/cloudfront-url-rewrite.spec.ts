import { App, Stack } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront'
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins'
import { describe, expect, it } from 'vitest'
import { CloudFrontUrlRewrite } from '../src/cloudfront-url-rewrite'

describe('CloudFrontUrlRewrite', () => {
	it('should create a CloudFront distribution with the correct behavior', () => {
		const app = new App()
		const stack = new Stack(app, 'TestStack')
		const distribution = new cloudfront.Distribution(stack, 'MyDistribution', {
			defaultBehavior: {
				origin: new origins.HttpOrigin('example.com'),
			},
		})

		new CloudFrontUrlRewrite(stack, 'MyUrlRewrite', {
			distribution,
			fromHostname: 'original.example.com',
			toHostname: 'new.example.com',
			redirectUriPattern: '^/oldpath/(.*)',
			targetUriPattern: '/newpath/$1',
		})

		const template = Template.fromStack(stack)
		expect(template.toJSON()).toMatchSnapshot()
	})
})
