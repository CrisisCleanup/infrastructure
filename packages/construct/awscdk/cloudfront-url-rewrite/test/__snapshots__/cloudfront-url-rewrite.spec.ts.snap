// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CloudFrontUrlRewrite > should create a CloudFront distribution with the correct behavior 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyDistribution6271DFB5": {
      "Properties": {
        "DistributionConfig": {
          "CacheBehaviors": [
            {
              "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
              "Compress": true,
              "FunctionAssociations": [
                {
                  "EventType": "viewer-request",
                  "FunctionARN": {
                    "Fn::GetAtt": [
                      "MyUrlRewriteHandlerC655E710",
                      "FunctionARN",
                    ],
                  },
                },
              ],
              "PathPattern": "/*",
              "TargetOriginId": "TestStackMyDistributionOrigin2B54A80B9",
              "ViewerProtocolPolicy": "allow-all",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "TestStackMyDistributionOrigin1485E8085",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": "example.com",
              "Id": "TestStackMyDistributionOrigin1485E8085",
            },
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": "original.example.com",
              "Id": "TestStackMyDistributionOrigin2B54A80B9",
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "MyUrlRewriteHandlerC655E710": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "// src/handler.function.ts
var FROM_HOSTNAME = \\"original.example.com\\";
var REDIRECT_URI_PATTERN = new RegExp(\\"^/oldpath/(.*)\\", \\"g\\");
var TO_HOSTNAME = \\"new.example.com\\";
var TARGET_URI_PATTERN = \\"/newpath/$1\\";
function handler(event) {
  var request = event.request;
  var headers = request.headers;
  var uri = request.uri;
  var hostParams = headers.host;
  var host = \\"\\";
  if (hostParams && hostParams.value) {
    host = hostParams.value;
  }
  if (!host || host !== FROM_HOSTNAME) {
    return request;
  }
  var newUri = uri.replace(REDIRECT_URI_PATTERN, TARGET_URI_PATTERN);
  var newUrl = \`https://\${TO_HOSTNAME}\${newUri}\`;
  return {
    statusCode: 301,
    statusDescription: \\"Moved Permanently\\",
    headers: {
      location: { value: newUrl }
    }
  };
}",
        "FunctionConfig": {
          "Comment": "src/handler.lambda.ts",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::Region",
              },
              "TestStackMyUrlRewriteHandlerD49D13ED",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFront::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
